content/blog/2022-03-24-e2e-computer-vision.md

```diff
@@ -0,0 +1,420 @@
+
+There is a super convenient set of
+[Experiment Management](https://dvc.org/doc/user-guide/experiment-management)
+features that make switching between reproducible experiments very easy without
+adding failed experiments to your git commit history and/or branches. If you
+have a local machine powerful enough to quickly train CV models, then definitely
+give Experiment Management features a try and check out this
+[blog post](https://dvc.org/blog/ml-experiment-versioning). Here, however, we
+are assuming that we'll be modifying the parameters of our experiments locally
+while the training process itself will happen on cloud-based resources like AWS
+EC2.
+
+### Leveraging Cloud Resources with CI/CD and CML
+
+If you ever wondered if there's an easy way to quickly and easily:
+
+1. provision a powerful VM in the cloud
+2. submit your ML training job to it
+3. get the results back
+4. automatically shut down the VM without having to worry about excessive cloud
+   bills
+
+then the answer is yes if you use [CML library](https://cml.dev/) in combination
+with CI/CD tools like [GitHub Actions](https://docs.github.com/en/actions) or
+[GitLab CI/CD](https://docs.gitlab.com/ee/ci/).
+
+We've configured three
+[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/master/.github/workflows)
+for GitHub Actions, each of which corresponds to a particular stage depending on
+the project's lifecycle we are in:
+
+1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/train-model.yaml)
+
+   In this stage, we'll be working on an experimentation git branch:
+   experimenting with data preprocessing, changing model architecture, tuning
+   hyperparameters, etc. Once we think our experiment is ready to be run, we'll
+   push our changes to a remote repository (in this case, GitHub). This push
+   will trigger a CI/CD job in GitHub Actions, which in turn will:
+
+   1. provision an EC2 virtual machine with a GPU in AWS
+   2. deploy our experiment branch to this machine
+   3. rerun the entire DVC pipeline
+   4. push files with metrics and other DVC artifacts back to GitHub
+
+   At this point, we can assess the results in DVC Studio and GitHub and decide
+   what things we want to change next.
+
+```mermaid
+flowchart TB
+A(Commit to exp branch) -->|Push changes| B("Exp CML workflow\n(training & reporting)")
+B --> |Reports,\\nmetrics,\\nplots| C("Check results.\nAre they good?")
+C --> |No\\nchange experiment parameters | A
+C -->|Yes| D(PR to dev branch)
+```
+
+1. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/master/.github/workflows/dev-train-upload-deploy.yaml)
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the dev branch. This would trigger a different CI/CD job
+   that will:
+
+   1. retrain the model with the new parameters
```

Comment:
If I understand the workflow correctly, dvc won't actually run the pipeline, right?

If this is True, it feels like this workflow should not launch a cloud instance nor try to run the pipeline at all, unless we are considering the possibility of someone pushing changes directly to develop?